Remove dtype from torch.Storage and use only torch.ByteStorage

kurtamohler opened the pull request:
Remove dtype tracking from Python Storage interface, remove all the different `<type>Storage` classes except for `ByteStorage`, and update serialization accordingly, while maintaining as much FC/BC as possible

Fixes #47442


* **THE SERIALIZATION FORMAT IS FULLY FC/BC.** We worked very hard to make sure this is the case. We will probably want to break FC at some point to make the serialization structure of tensors make more sense, but not today.
* There is now only a single torch.ByteStorage class. Methods like `Tensor.set_` no longer check that the dtype of storage is appropriate. 
* As we no longer know what dtype of a storage is, we've **removed** the size method from Storage, replacing it with nbytes. This is to help catch otherwise silent errors where you confuse number of elements with number of bytes.
* `Storage._new_shared` takes a `nbytes` kwarg and will reject previous positional only calls.  `Storage._new_with_file` and `_set_from_file` require explicit element size arguments.
* It's no longer possible to convert storages to different types using the float/double/etc methods. Instead, do the conversion using a tensor.
* It's no longer possible to allocate a typed storage directly using FloatStorage/DoubleStorage/etc constructors. Instead, construct a tensor and extract its storage. The classes still exist but they are used purely for unpickling.
* The preexisting serialization format stores dtype with storage, and in fact this dtype is used to determine the dtype of the tensor overall. 
 To accommodate this case, we introduce a new TypedStorage concept that exists only during unpickling time which is used to temporarily store the dtype so we can construct a tensor. **If you overrode the handling of pickling/unpickling, you MUST add handling for TypedStorage** or your serialization code will degrade to standard file-based serialization.

kurtamohler committed:
Remove dtype from torch.Storage and use only torch.ByteStorage

vadimkantorov commented:
I think I used other storages for conversion of Python's bytearray to a given dtype. Related: https://github.com/pytorch/pytorch/issues/30387 https://github.com/pytorch/pytorch/issues/29013

kurtamohler commented:
@vadimkantorov, I think it should be possible to do that after dtypes are removed from storage. You would need to load the bytearray into a storage and then create a tensor of the correct dtype to view the storage.

kurtamohler commented:
Hey @ezyang, I'm getting failures with sharing memory between processes. I'm not very familiar yet with how PyTorch sets up shared memory, so I figured I'd ping you to ask if there are any obvious places in the codebase I should look.

Tests are showing that sharing is not happening. Mutating a tensor that is supposed to be shared only changes it in one process, not both.  The data pointers match for both the tensor and the storage (`tensor.data_ptr()` and `tensor.storage()._cdata`, respectively) but evidently the virtual addresses are not actually resolving to the same physical location in the memory maps of the two processes. I haven't actually verified that by looking at the memory maps of the processes, but that seems like the only possible explanation. CUDA shared memory is evidently working fine, just CPU is broken.

ezyang commented:
ForkingPickler is for transient pickles in multiprocessing, FC shouldn't be necessary!

ezyang commented:
Anything I can do to help move this along? :)

kurtamohler commented:
Hey @ezyang, sorry for the delay. I was meaning to give an update about where I am, and I'd appreciate your thoughts. There are a few things I have left to do:

* Test serialization forward compatibility. I guess that I'll have to test this locally by running a script that saves tensors in a bunch of different ways (regular save, legacy save, shared storages (only if dtypes are the same), package exporter, JIT, ...), and then try loading them using an old version of PyTorch
* Make sure BC tests are thorough
* A couple small TODOs remaining (or at least remove TODO comments :P)
* Update documentation to reflect all of these changes
* I should maybe try to break this into a few stacked PRs. I think a lot of the changes to serialization are separable and can be added first. Then actually removing dtype from Storage would go in last. Do you think that's a good idea? On the one hand, it might slow things down a little bit (deciding which changes can be encapsulated into one PR, as well as any intermediate changes that would need to be made before Storage is actually untyped), but it should also make it simpler to make sure we're not breaking anything.
* Add tests for new functionality that wasn't previously possible. For instance, saving and loading tensors of different dtypes that share the same storage.


I'd also like to summarize the user-facing changes here and make sure we agree on them:

* At the moment, `torch.ByteStorage` and `torch.cuda.ByteStorage` are the only instantiable storage class. We also still have `torch.Storage`, which will always translate to `torch.ByteStorage` now. Would it be better to rename `ByteStorage` to just `Storage`? I feel like there's no harm in keeping both for now.
* Remove methods for converting storages to different types
* `Tensor.set_(storage)` no longer checks storage's type
* I changed the behavior of `torch.tensor(storage, dtype=dtype)`. Before, this call would cast each element of the storage to the desired dtype. But that no longer makes much sense. Say we have a ByteStorage that we are viewing as torch.float64, and we want to view it as torch.complex128 instead. We wouldn't want to cast each byte of the storage to its own complex number. So instead, `torch.tensor(storage, dtype=dtype)` doesn't do any casting any more, it just sets the output tensor's dtype and size appropriately to view the provided storage.
* After saving a Storage object directly (ie, not as part of a tensor) and loading it back up, it will be wrapped in a `torch.storage.TypedStorage`. Of course, we can only save `torch.ByteStorage` now. But we can also load other types of storages if they were saved in an older version of PyTorch--in this case they will be loaded as a `torch.ByteStorage`, but paired with the appropriate `torch.dtype` inside a `torch.storage.TypedStorage`.
* Saving a list like `[tensor.storage(), tensor]` is not FC. If we try to load it in an older PyTorch, the dtype of the storage will always be `ByteStorage`, but the tensor can be any dtype. In master PyTorch, we get an error if we try to assign a storage to a tensor of different types.
* We should have FC and BC for everything (including serialization) that does not manipulate storage objects directly (like if we're only dealing with tensors). Although as I mentioned above, I still need to thoroughly test FC and make sure BC tests are good

I *think* that's pretty much everything. Let me know your thoughts

ezyang commented:
> I should maybe try to break this into a few stacked PRs. I think a lot of the changes to serialization are separable and can be added first. Then actually removing dtype from Storage would go in last. Do you think that's a good idea? On the one hand, it might slow things down a little bit (deciding which changes can be encapsulated into one PR, as well as any intermediate changes that would need to be made before Storage is actually untyped), but multiple PRs should also make it easier to make sure we're not breaking anything.

I skimmed the PR and I don't think this will be necessary yet. If you can give an order to review files in that will be a big help. (We might split out easy to split out stuff later if we change our minds).

ezyang commented:
> Would it be better to rename ByteStorage to just Storage? I feel like there's no harm in keeping both for now.

I think torch.Storage is better but this can wait. 

The API changes all look good. If you want to be SUPER nice to legacy users, we probably would teach some APIs like `set_` to know how to take `TypedStorage`, but I'm willing to wait for someone to holler in pain before you do that.

I guess I can start reviewing the substantive changes then? It sounds like everything else is window dressing.

kurtamohler commented:
> I guess I can start reviewing the substantive changes then?

Yeah I think that's probably a good idea, thanks!

ezyang commented on torch/_utils.py:
```diff
@@ -185,6 +185,7 @@ def _rebuild_mlc_tensor(data, dtype, device, requires_grad):
     return tensor
 
 
+# TODO: This will need to be fixed to work with TypedStorage
```
Did you fix this?

ezyang commented on torch/_utils.py:
```diff
@@ -513,3 +514,21 @@ def _handle_complex(tensor):
+        # NOTE: torch.bool is not supported in torch.iinfo()
+        return 1
+    else:
+        return torch.iinfo(dtype).bits >> 3
```
Huh, it's weird that we never had this already

kurtamohler replied:
Yeah I thought so too. It's possible that we do have a similar function elsewhere, but I couldn't find it

ezyang commented on torch/csrc/generic/Storage.cpp:
```diff
@@ -319,7 +319,6 @@ typedef PyObject *(*getter)(PyObject *, void *);
 // NOLINTNEXTLINE(cppcoreguidelines-avoid-c-arrays,modernize-avoid-c-arrays,cppcoreguidelines-avoid-non-const-global-variables)
 static struct PyGetSetDef THPStorage_(properties)[] = {
   {"device", (getter)THPStorage_(device), nullptr, nullptr, nullptr},
-  {"dtype",  (getter)THPStorage_(dtype), nullptr, nullptr, nullptr},
```
This can be later, but one consequence of untyped Storage is I hope we can remove Storage.cpp and related files out of generic/ (which is still doing old style TH macro substitution)

ezyang commented on torch/_utils.py:
```diff
@@ -129,8 +129,8 @@ def _get_async_or_non_blocking(function_name, non_blocking, kwargs):
-    t = torch.tensor([], dtype=storage.dtype, device=storage.device)
-    return t.set_(storage, storage_offset, size, stride)
+    t = torch.tensor([], dtype=storage.dtype, device=storage.storage.device)
+    return t.set_(storage.storage, storage_offset, size, stride)
```
TODO: double check this later

ezyang commented on torch/csrc/generic/StorageMethods.cpp:
```diff
@@ -275,12 +285,20 @@ static PyObject *THPStorage_(setFromFile)(PyObject *_self, PyObject *args)
   bool is_real_file = PyTuple_GET_ITEM(args, 2) == Py_True;
 
+  // TODO: Is there any other way to get the correct element size
+  // without adding a new arg? I don't think so, but let's make sure
```
Seems unlikely

ezyang commented on torch/csrc/generic/serialization.cpp:
```diff
@@ -99,48 +99,52 @@ THWStorage * THPStorage_(readFileRaw)(io file, THWStorage *_storage)
   data = (scalar_t*)cpu_data.get();
 #endif
 
+  // TODO: Change sizeof(scalar_t) to be element_size
```
you still gonna do this?

ezyang commented on torch/csrc/utils/tensor_new.cpp:
```diff
@@ -91,12 +91,16 @@ Tensor new_with_tensor(c10::TensorOptions options, at::ScalarType scalar_type, c
     if (length < 0) throw python_error();
+    if (is_storage) {
+      length /= elementSize(scalar_type);
+    }
```
oof

ezyang commented on torch/multiprocessing/reductions.py:
```diff
@@ -99,7 +99,7 @@ def rebuild_tensor(cls, storage, metadata):
 
 def rebuild_cuda_tensor(tensor_cls, tensor_size, tensor_stride, tensor_offset,
-                        storage_cls, storage_device, storage_handle, storage_size_bytes, storage_offset_bytes,
+                        storage_cls, dtype, storage_device, storage_handle, storage_size_bytes, storage_offset_bytes,
```
This looks dodgy, for BC/FC.

kurtamohler replied:
This is only used in ForkingPickler, so it should be ok, right?

ezyang replied:
oops, that's right! you're clear

ezyang commented on torch/package/_directory_reader.py:
```diff
@@ -44,9 +32,10 @@ def get_record(self, name):
-        return _HasStorage(storage.from_file(filename=filename, size=numel))
+        nbytes = torch._utils._element_size(dtype) * numel
+        storage = cast(Storage, torch.ByteStorage)
+        return _HasStorage(storage.from_file(filename=filename, size=nbytes))
```
It maybe marginally safer to rename the kwarg here to nbytes (from size), to help other people catch if they were feeding numel rather than bytes.

ezyang commented on torch/serialization.py:
```diff
@@ -406,28 +407,74 @@ def persistent_id(obj: Any) -> Optional[Tuple]:
             return ('module', obj, source_file, source)
 
-        elif torch.is_storage(obj):
+        if isinstance(obj, torch.storage.TypedStorage) or torch.is_storage(obj):
```
Something that will be helpful is to add notes about when code can be deleted (and how to do the deletion), so people don't perpetually tiptoe around TypedStorage

kurtamohler replied:
I've added some TODOs where we should no longer need to use TypedStorage when we decide to break FC

ezyang commented on torch/serialization.py:
```diff
@@ -406,28 +407,74 @@ def persistent_id(obj: Any) -> Optional[Tuple]:
+            # expect. `tensor` will be able to load up properly in old
+            # versions, pointing to a FloatStorage. However, `storage` is still
+            # being translated to a ByteStorage, and it will try to resolve to
+            # the same FoatStorage that `tensor` contains. This will also cause
```
FloatStorage

ezyang commented on torch/utils/data/_utils/collate.py:
```diff
@@ -51,7 +51,7 @@ def default_collate(batch):
             # shared memory tensor to avoid an extra copy
             numel = sum([x.numel() for x in batch])
-            storage = elem.storage()._new_shared(numel)
+            storage = elem.storage()._new_shared(numel * elem.element_size())
```
Also wondering if we can kwarg this one, to help catch other sites where they are still passing numel

ezyang commented on test/package/test_torchscript.py:
```diff
@@ -412,13 +421,13 @@ def test_load_shared_tensors(self):
         importer = PackageImporter(buffer)
         loaded_mod_1 = importer.load_pickle("res", "mod1.pkl")
 
-        self.assertTrue(
```
lol how did these tests ever work

kurtamohler replied:
Haha yeah, I guess `_cdata` was getting cast to a bool, and since it's always going to be nonzero, it's True

ezyang commented on test/test_cuda.py:
```diff
@@ -580,9 +580,10 @@ def test_serialization_array_with_storage(self):
         self.assertTrue(isinstance(q_copy[2], torch.cuda.FloatTensor))
-        self.assertTrue(isinstance(q_copy[3], torch.cuda.IntStorage))
+        self.assertTrue(isinstance(q_copy[3], torch.storage.TypedStorage))
+        self.assertTrue(isinstance(q_copy[3].storage, torch.cuda.ByteStorage))
```
Is this the kind of test where y and y.storage() were stored in the same place, and it caused you the FC headaches (which you wrote a note about?) I wonder if there should be some sort of adjustment to the test itself, so it isn't trying to serialize both y and its storage at the same time, in one data structure.  OTOH this works just fine if you don't go backwards in time, so maybe not.

kurtamohler replied:
Right, this is a non-FC case. But yeah it still works even though it isn't FC, so I think we should keep testing it

ezyang commented on test/test_jit.py:
```diff
@@ -10154,7 +10154,7 @@ def foo(self):
                 m_import = self.getExportImportCopy(m_orig)
                 # check to make sure the storage wasn't resized
-                self.assertTrue(m_orig.param.storage().size() == 25)
+                self.assertTrue(m_orig.param.storage().size() == 25 * m_orig.param.element_size())
```
Oof. This one will be difficult to warn against, without just removing the size() method entirely from storage(). Hopefully not too many people affected by it.

kurtamohler replied:
Hmm, well we could just rename it to `nbytes`

ezyang replied:
yeah let's rename it

ezyang commented:
Going to need detailed "how to migrate your code" notes for the release notes. Test suite is a good way to get started writing.

ezyang commented on test/test_torch.py:
```diff
@@ -6965,14 +6899,18 @@ def _get_tensors(**kwargs):
         for device in devices:
             x = torch.tensor([], device=device)
-            self.assertEqual(x.dtype, x.storage().dtype)
+            # TODO: Shoud we even have this test anymore?
```
I'd kill it

ezyang commented on test/test_vmap.py:
```diff
@@ -1770,20 +1770,20 @@ def test_new_empty_strided(self):
             x = torch.randn(B0)
             result = vmap(lambda x: x.new_empty_strided(size, stride))(x)
-            S = torch.empty_strided(size, stride).storage().size()
+            S = int(torch.empty_strided(size, stride).storage().size() / x.element_size())
```
use integer division here instead of manual cast to int after float div?

ezyang commented on torch/testing/_internal/common_utils.py:
```diff
@@ -2315,3 +2316,42 @@ def disable_gc():
         yield
+
+# Returns scalar tensor representation of a list of integer byte values
+def bytes_to_scalar(byte_list: List[int], dtype: torch.dtype, device: torch.device):
```
Wondering if you need to unit test this. I guess this is being cross tested with Tensor storage so not a big deal

kurtamohler replied:
Test added

ezyang commented:
Overall: this is a marvel. I'm so impressed how far this PR goes. Really good work!

I have done a quick review of every file except torch/serialization.py. Planning to do that tomorrow. Mostly everything looks good, though the CUDA FC/BC situation looks a little dicey. Release notes definitely needed for this PR.

@Lilyjjo since you are interested in serialization, I think you should probably review this PR too, and ask any questions clarifications you might have.

kurtamohler commented:
Thanks for the review @ezyang! I've made your requested changes, except for migration help notes. I still have to do most of the things I listed in my comment above

kurtamohler commented:
I wrote a standalone test script for checking serialization FC/BC: https://github.com/kurtamohler/pytorch-perf-test-scripts/blob/master/untyped-Storage/serialization_compat_test.py

It doesn't test everything I want it to yet. Right now, it only tests regular serialization (like `torch.save/load` and `pickle.dump/load`). It tests multiple different pickle protocols, different dtypes, CPU/CUDA, storage sharing, both `pickle` and `dill`. But it doesn't yet test serialization for JIT, packaging, dataloader, etc. For the stuff it does test, this PR passes FC and BC so far.

dreiss commented:
> But it doesn't yet test serialization for JIT

I would suggest at least a very simple JIT test (trace script a single Linear module), since I believe it uses a totally different code path compared to `torch.save`.

ezyang commented on torch/storage.py:
```diff
@@ -146,20 +101,91 @@ def share_memory_(self):
+    if not hasattr(_pickle_storage_type_to_dtype_map, 'cache'):
+        _pickle_storage_type_to_dtype_map.cache = {  # type: ignore[attr-defined]
+            val: key for key, val in _dtype_to_pickle_storage_type_map().items()}
+    return _pickle_storage_type_to_dtype_map.cache  # type: ignore[attr-defined]
```
A more compact way to write this is just `@lru_cache`

kurtamohler replied:
Oh cool, didn't know about that!

ezyang commented on torch/serialization.py:
```diff
@@ -406,28 +407,75 @@ def persistent_id(obj: Any) -> Optional[Tuple]:
             view_metadata: Optional[Tuple[str, int, int]]
-            obj = cast(Storage, obj)
-            storage_type = normalize_storage_type(type(obj))
+            storage = cast(Storage, storage)
```
instead of casting here, it seems better to `assert isinstance(obj, Storage)` in else branch and then mypy should be able to figure it out

ezyang commented on torch/serialization.py:
```diff
@@ -406,28 +407,75 @@ def persistent_id(obj: Any) -> Optional[Tuple]:
+
+            if storage_key not in serialized_storages:
+                serialized_storages[storage_key] = (storage, dtype)
+            is_view = storage._cdata != storage._cdata
```
yuck (but I see this is preexisting code)

ezyang commented on torch/serialization.py:
```diff
@@ -690,14 +764,23 @@ def persistent_load(saved_id):
                     key, location, storage_type = args
-                    obj = storage_type._new_with_file(f)
+                    dtype = torch.storage._get_dtype_from_pickle_storage_type(storage_type.__name__)
+                    obj = cast(Storage, torch.ByteStorage)._new_with_file(f, torch._utils._element_size(dtype))
```
huh, surprised you need the cast here

ezyang commented on torch/serialization.py:
```diff
@@ -706,18 +789,18 @@ def persistent_load(saved_id):
                     stride = struct.unpack(f'<{ndim}q', f.read(8 * ndim))
                     storage_offset, = struct.unpack('<q', f.read(8))
-                    tensor = tensor_type().set_(storage, storage_offset, size, stride)
+                    tensor = torch.tensor([], dtype=storage.dtype).set_(
```
nit: it would be marginally more efficient here to do torch.empty(0, dtype=storage.dtype) as it will skip parsing the Python [] to figure out what contents to put in.

ezyang commented:
OK, well, this is still marked as WIP, but I'm pretty ready to pull the trigger on this one and see how it goes. Earlier the better, so it has more time to bake on master. More testing always good, so don't feel rushed on that.

ezyang commented:
test failures are real but look easy to fix

```
Traceback (most recent call last):
  File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 397, in instantiated_test
    result = test_fn(self, *args)
  File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 876, in only_fn
    return fn(slf, device, *args, **kwargs)
  File "test_numpy_interop.py", line 368, in test_parse_numpy_int
    self.assertEqual(torch.Storage(np_val).size(), scalar)  # type: ignore[attr-defined]
AttributeError: 'torch.ByteStorage' object has no attribute 'size'
		
```

kurtamohler commented:
@ezyang, alright I'm good with that if you are! I'll make followup PR(s) for the remaining improvements

kurtamohler commented:
I've added a few more test cases to my serialization compatibility test script (https://github.com/kurtamohler/pytorch-perf-test-scripts/blob/master/untyped-Storage/serialization_compat_test.py), including JIT script/tracing and quantization. They're all looking good--so far no BC/FC issues.

However, it looks like my PR probably has some significant performance issues. When I run `serialization_compat_test.py --load` on master, it takes about 5 seconds, but if I run it on this branch, it takes about 12 seconds. I don't know yet what's causing the slowdown, though it wouldn't surprise me if some of the division operations I added are part of it. Unfortunately, some of those might not be avoidable. In any case, I'll do some profiling to find the cause.

ezyang commented:
need rebase

peterbell10 commented on torch/_storage_docs.py:
```diff
@@ -30,19 +20,18 @@ def add_docstr_all(method, docstr):
-then the file must contain at least `size * sizeof(Type)` bytes
-(`Type` is the type of storage). If `shared` is `True` the file will be
-created if needed.
+`nbutes` is the number of bytes in the storage. If `shared` is `False`, then
```
```suggestion
`nbytes` is the number of bytes in the storage. If `shared` is `False`, then
```

kurtamohler commented:
@ezyang, good news--the performance issue I was seeing was just coming from the fact that I was using `storage.tolist()` to compare storages. Since storages are typed on master, there are fewer values to compare. If I remove these comparisons, the performance matches

kurtamohler committed:
Fix typos

kurtamohler committed:
Merge branch 'master' of github.com:kurtamohler/pytorch into storage-remove-dtype-47442

kurtamohler committed:
Use same device

kurtamohler committed:
Merge branch 'master' of github.com:kurtamohler/pytorch into storage-remove-dtype-47442

ezyang commented:
need to handle the BC breaking changes in fbcode (very thankful for this, easy to find sites since we did bc break them, otherwise would be silent correctness)

ezyang commented:
@kurtamohler How difficult would it be to restore support for saying torch.FloatStorage(...)? This currently raises an error but it would fix a couple tests directly if this usage pattern still worked.

ezyang commented:
Substantial internal roadbump: there's an internal copy paste of torch/serialization.py (pytorch/manifold/serialization.py) that I need to port the serialization.py changes here to (which in practice means I need to take the latest version of serialization.py as written here and then reapply the changes to D18525979 and all subsequent diffs (only five, mercifully). This will take a little time.

cc @lw

kurtamohler commented:
> How difficult would it be to restore support for saying torch.FloatStorage(...)?

Probably not overly difficult. I assume we still would not want to instantiate a `FloatStorage` though, right? We could add `torch.FloatStorage()` back in but make it return a `ByteStorage` wrapped in a `TypedStorage`.

ezyang commented:
@kurtamohler if you want to go slow on that request for now, it's OK; it seems that all of these errors are coming from a single place which I need to fix differently

ezyang commented:
I just fixed a fascinating internal bug, where the fix diff is this:

```
@@ -30,6 +30,7 @@
 
         for t in torch._storage_classes:
             self._dispatch_table[t] = TorchMpReductions.reduce_storage
+        self._dispatch_table[torch.storage.TypedStorage] = TorchMpReductions.reduce_typed_storage
 
         for t in torch._tensor_classes:
             self._dispatch_table[t] = TorchMpReductions.reduce_tensor
```

This is some internal reimplementation of multiprocessing reductions, and it caused test failures in a very subtle way: without the dispatch table override for TypedStorage, the TypedStorage reduction was done using the conventional file system serialization, which meant that tensors were no longer shared across processes. Luckily, this test was explicitly testing sharing was occurring, but if there are other reimplementations of multiprocessing reductions we are probably not going to be as lucky.

I'm wondering if there is any way we can get this case to error. I'm a bit doubtful... there is just no logical interposition point. Another one for the release notes, I suppose.

kurtamohler commented:
Thanks for fixing it @ezyang! It's not obvious to me how we would make similar cases error

ezyang commented:
Only six more failing internal projects, incredible.

kurtamohler committed:
Merge branch 'master' into storage-remove-dtype-47442

kurtamohler committed:
Merge branch 'master' into storage-remove-dtype-47442

ezyang commented:
btw, next time, ghstack diff would be preferred, easier to keep fbcode changes this way

ezyang commented:
Another internal failure I fixed looked like this:

```
values = ", ".join(str(element) for element in v.storage())
```

I fixed it by changing it to

```
values = ", ".join(str(element) for element in v.resize(-1).tolist())
```

But maybe we should have also disabled iterating on storage...

ezyang committed:
Merge remote-tracking branch 'origin/master' into storage-remove-dtype-47442

kurtamohler commented:
> But maybe we should have also disabled iterating on storage...

We could do that if you'd like. I think it would still need to at least have indexing

ezyang commented:
> We could do that if you'd like. I think it would still need to at least have indexing

The rule I'm applying here is "no code which previously worked should silently give an incorrect result in the new version." This implies indexing as `x.storage()[1]` should stop working as it would indeed give a silently incorrect result. Similar with `resize_()`. If we're going to do this properly we probably need to audit all of the user visible storage API for this property.

ezyang commented:
I'm unlinking this PR from the diff, to open a new PR using ghstack, so that I can more easily ship changes to and from fbcode.

ezyang commented:
Please push changes to https://github.com/pytorch/pytorch/pull/62030 ; I've also resolved the merged conflicts

kurtamohler commented:
Closing this, since it was moved to #62030 a while ago
